### ISSUE TITLE
HTTP Continuation

### ISSUE THREAD

**Issue by inline-github** [warmcat/libwebsockets#2483]
Hello all,

This issue is rarely reproducible so only pcap trace is available 
I am using client libwebsockets v2.3-stable release. 

Unfortunately it was captured pcap trace only in one direction 
Let me describe pcap trace observation:
1. client libwebsockets sent request to establish connection on 800 port to WebSocket server.
93149	8375.003820	192.168.160.3	192.168.160.2	TCP	74	46450 → 800 [ACK] Seq=80333 Ack=1059200 Win=182272 Len=0 TSval=768628 TSecr=3921104
Frame 93149: 74 bytes on wire (592 bits), 74 bytes captured (592 bits)
Ethernet II, Src: 12:20:30:40:50:65 (12:20:30:40:50:65), Dst: 06:01:02:00:00:00 (06:01:02:00:00:00)
802.1Q Virtual LAN, PRI: 0, DEI: 0, ID: 1
Internet Protocol Version 4, Src: 192.168.160.3, Dst: 192.168.160.2
Transmission Control Protocol, Src Port: 46450, Dst Port: 800, Seq: 80333, Ack: 1059200, Len: 0
    Source Port: 46450
    Destination Port: 800
    [Stream index: 383]
    [TCP Segment Len: 0]
    Sequence Number: 80333    (relative sequence number)
    Sequence Number (raw): 4278264696
    [Next Sequence Number: 80333    (relative sequence number)]
    Acknowledgment Number: 1059200    (relative ack number)
    Acknowledgment number (raw): 3375712623
    1000 .... = Header Length: 32 bytes (8)
    Flags: 0x010 (ACK)
        000. .... .... = Reserved: Not set
        ...0 .... .... = Nonce: Not set
        .... 0... .... = Congestion Window Reduced (CWR): Not set
        .... .0.. .... = ECN-Echo: Not set
        .... ..0. .... = Urgent: Not set
        .... ...1 .... = Acknowledgment: Set
        .... .... 0... = Push: Not set
        .... .... .0.. = Reset: Not set
        .... .... ..0. = Syn: Not set
        .... .... ...0 = Fin: Not set
        [TCP Flags: ·······A····]
    Window: 1424
    [Calculated window size: 182272]
    [Window size scaling factor: 128]
    Checksum: 0x3299 [unverified]
    [Checksum Status: Unverified]
    Urgent Pointer: 0
    Options: (12 bytes), No-Operation (NOP), No-Operation (NOP), Timestamps
    [Timestamps]

2. client libwebsockets has sent ACK on received SYN, ACK from WebSocket server 
93151	8375.130499	192.168.160.3	192.168.160.2	TCP	74	46614 → 800 [ACK] Seq=1 Ack=1 Win=29312 Len=0 TSval=768639 TSecr=3922241
Frame 93151: 74 bytes on wire (592 bits), 74 bytes captured (592 bits)
Ethernet II, Src: 12:20:30:40:50:65 (12:20:30:40:50:65), Dst: 06:01:02:00:00:00 (06:01:02:00:00:00)
802.1Q Virtual LAN, PRI: 0, DEI: 0, ID: 1
Internet Protocol Version 4, Src: 192.168.160.3, Dst: 192.168.160.2
Transmission Control Protocol, Src Port: 46614, Dst Port: 800, Seq: 1, Ack: 1, Len: 0
    Source Port: 46614
    Destination Port: 800
    [Stream index: 384]
    [TCP Segment Len: 0]
    Sequence Number: 1    (relative sequence number)
    Sequence Number (raw): 1946237506
    [Next Sequence Number: 1    (relative sequence number)]
    Acknowledgment Number: 1    (relative ack number)
    Acknowledgment number (raw): 4141163240
    1000 .... = Header Length: 32 bytes (8)
    Flags: 0x010 (ACK)
        000. .... .... = Reserved: Not set
        ...0 .... .... = Nonce: Not set
        .... 0... .... = Congestion Window Reduced (CWR): Not set
        .... .0.. .... = ECN-Echo: Not set
        .... ..0. .... = Urgent: Not set
        .... ...1 .... = Acknowledgment: Set
        .... .... 0... = Push: Not set
        .... .... .0.. = Reset: Not set
        .... .... ..0. = Syn: Not set
        .... .... ...0 = Fin: Not set
        [TCP Flags: ·······A····]
    Window: 229
    [Calculated window size: 29312]
    [Window size scaling factor: 128]
    Checksum: 0x9f41 [unverified]
    [Checksum Status: Unverified]
    Urgent Pointer: 0
    Options: (12 bytes), No-Operation (NOP), No-Operation (NOP), Timestamps
    [Timestamps]

3. client libwebsockets sent GET HTTP request to WebSocket server
93152	8375.130731	192.168.160.3	192.168.160.2	HTTP	405	GET / HTTP/1.1 
Frame 93152: 405 bytes on wire (3240 bits), 405 bytes captured (3240 bits)
Ethernet II, Src: 12:20:30:40:50:65 (12:20:30:40:50:65), Dst: 06:01:02:00:00:00 (06:01:02:00:00:00)
802.1Q Virtual LAN, PRI: 0, DEI: 0, ID: 1
Internet Protocol Version 4, Src: 192.168.160.3, Dst: 192.168.160.2
Transmission Control Protocol, Src Port: 46614, Dst Port: 800, Seq: 1, Ack: 1, Len: 331
    Source Port: 46614
    Destination Port: 800
    [Stream index: 384]
    [TCP Segment Len: 331]
    Sequence Number: 1    (relative sequence number)
    Sequence Number (raw): 1946237506
    [Next Sequence Number: 332    (relative sequence number)]
    Acknowledgment Number: 1    (relative ack number)
    Acknowledgment number (raw): 4141163240
    1000 .... = Header Length: 32 bytes (8)
    Flags: 0x018 (PSH, ACK)
        000. .... .... = Reserved: Not set
        ...0 .... .... = Nonce: Not set
        .... 0... .... = Congestion Window Reduced (CWR): Not set
        .... .0.. .... = ECN-Echo: Not set
        .... ..0. .... = Urgent: Not set
        .... ...1 .... = Acknowledgment: Set
        .... .... 1... = Push: Set
        .... .... .0.. = Reset: Not set
        .... .... ..0. = Syn: Not set
        .... .... ...0 = Fin: Not set
        [TCP Flags: ·······AP···]
    Window: 229
    [Calculated window size: 29312]
    [Window size scaling factor: 128]
    Checksum: 0x34f7 [unverified]
    [Checksum Status: Unverified]
    Urgent Pointer: 0
    Options: (12 bytes), No-Operation (NOP), No-Operation (NOP), Timestamps
    [SEQ/ACK analysis]
    [Timestamps]
    TCP payload (331 bytes)
Hypertext Transfer Protocol
    GET / HTTP/1.1\r\n
    Pragma: no-cache\r\n
    Cache-Control: no-cache\r\n
    Host: 192.168.160.2:800\r\n
    Origin: http://192.168.160.2:800\r\n
    Upgrade: websocket\r\n
    Connection: Upgrade\r\n
    Sec-WebSocket-Key: bqaHIIM3QT8aHdi516hJFQ==\r\n
    Sec-WebSocket-Extensions: permessage-deflate; client_no_context_takeover; client_max_window_bits\r\n
    Sec-WebSocket-Version: 13\r\n
    \r\n
    [Full request URI: http://192.168.160.2:800/]
    [HTTP request 1/1]

4. client libwebsockets has sent ACK to WebSocket server
93153	8375.139069	192.168.160.3	192.168.160.2	TCP	74	46614 → 800 [ACK] Seq=332 Ack=130 Win=30336 Len=0 TSval=768640 TSecr=3922245
Frame 93153: 74 bytes on wire (592 bits), 74 bytes captured (592 bits)
Ethernet II, Src: 12:20:30:40:50:65 (12:20:30:40:50:65), Dst: 06:01:02:00:00:00 (06:01:02:00:00:00)
802.1Q Virtual LAN, PRI: 0, DEI: 0, ID: 1
Internet Protocol Version 4, Src: 192.168.160.3, Dst: 192.168.160.2
Transmission Control Protocol, Src Port: 46614, Dst Port: 800, Seq: 332, Ack: 130, Len: 0
    Source Port: 46614
    Destination Port: 800
    [Stream index: 384]
    [TCP Segment Len: 0]
    Sequence Number: 332    (relative sequence number)
    Sequence Number (raw): 1946237837
    [Next Sequence Number: 332    (relative sequence number)]
    Acknowledgment Number: 130    (relative ack number)
    Acknowledgment number (raw): 4141163369
    1000 .... = Header Length: 32 bytes (8)
    Flags: 0x010 (ACK)
        000. .... .... = Reserved: Not set
        ...0 .... .... = Nonce: Not set
        .... 0... .... = Congestion Window Reduced (CWR): Not set
        .... .0.. .... = ECN-Echo: Not set
        .... ..0. .... = Urgent: Not set
        .... ...1 .... = Acknowledgment: Set
        .... .... 0... = Push: Not set
        .... .... .0.. = Reset: Not set
        .... .... ..0. = Syn: Not set
        .... .... ...0 = Fin: Not set
        [TCP Flags: ·······A····]
    Window: 237
    [Calculated window size: 30336]
    [Window size scaling factor: 128]
    Checksum: 0x9d68 [unverified]
    [Checksum Status: Unverified]
    Urgent Pointer: 0
    Options: (12 bytes), No-Operation (NOP), No-Operation (NOP), Timestamps
    [Timestamps]

5. client libwebsockets has sent HTTP Continuation to WebSocket server
93154	8375.139956	192.168.160.3	192.168.160.2	HTTP	184	Continuation
Frame 93154: 184 bytes on wire (1472 bits), 184 bytes captured (1472 bits)
Ethernet II, Src: 12:20:30:40:50:65 (12:20:30:40:50:65), Dst: 06:01:02:00:00:00 (06:01:02:00:00:00)
802.1Q Virtual LAN, PRI: 0, DEI: 0, ID: 1
Internet Protocol Version 4, Src: 192.168.160.3, Dst: 192.168.160.2
Transmission Control Protocol, Src Port: 46614, Dst Port: 800, Seq: 332, Ack: 130, Len: 110
    Source Port: 46614
    Destination Port: 800
    [Stream index: 384]
    [TCP Segment Len: 110]
    Sequence Number: 332    (relative sequence number)
    Sequence Number (raw): 1946237837
    [Next Sequence Number: 442    (relative sequence number)]
    Acknowledgment Number: 130    (relative ack number)
    Acknowledgment number (raw): 4141163369
    1000 .... = Header Length: 32 bytes (8)
    Flags: 0x018 (PSH, ACK)
        000. .... .... = Reserved: Not set
        ...0 .... .... = Nonce: Not set
        .... 0... .... = Congestion Window Reduced (CWR): Not set
        .... .0.. .... = ECN-Echo: Not set
        .... ..0. .... = Urgent: Not set
        .... ...1 .... = Acknowledgment: Set
        .... .... 1... = Push: Set
        .... .... .0.. = Reset: Not set
        .... .... ..0. = Syn: Not set
        .... .... ...0 = Fin: Not set
        [TCP Flags: ·······AP···]
    Window: 237
    [Calculated window size: 30336]
    [Window size scaling factor: 128]
    Checksum: 0xd27a [unverified]
    [Checksum Status: Unverified]
    Urgent Pointer: 0
    Options: (12 bytes), No-Operation (NOP), No-Operation (NOP), Timestamps
    [SEQ/ACK analysis]
    [Timestamps]
    TCP payload (110 bytes)
Hypertext Transfer Protocol
    Data (110 bytes)

6. client libwebsockets has sent HTTP Continuation to WebSocket server
93155	8375.154377	192.168.160.3	192.168.160.2	HTTP	168	Continuation
Frame 93155: 168 bytes on wire (1344 bits), 168 bytes captured (1344 bits)
Ethernet II, Src: 12:20:30:40:50:65 (12:20:30:40:50:65), Dst: 06:01:02:00:00:00 (06:01:02:00:00:00)
802.1Q Virtual LAN, PRI: 0, DEI: 0, ID: 1
Internet Protocol Version 4, Src: 192.168.160.3, Dst: 192.168.160.2
Transmission Control Protocol, Src Port: 46614, Dst Port: 800, Seq: 442, Ack: 334, Len: 94
    Source Port: 46614
    Destination Port: 800
    [Stream index: 384]
    [TCP Segment Len: 94]
    Sequence Number: 442    (relative sequence number)
    Sequence Number (raw): 1946237947
    [Next Sequence Number: 536    (relative sequence number)]
    Acknowledgment Number: 334    (relative ack number)
    Acknowledgment number (raw): 4141163573
    1000 .... = Header Length: 32 bytes (8)
    Flags: 0x018 (PSH, ACK)
        000. .... .... = Reserved: Not set
        ...0 .... .... = Nonce: Not set
        .... 0... .... = Congestion Window Reduced (CWR): Not set
        .... .0.. .... = ECN-Echo: Not set
        .... ..0. .... = Urgent: Not set
        .... ...1 .... = Acknowledgment: Set
        .... .... 1... = Push: Set
        .... .... .0.. = Reset: Not set
        .... .... ..0. = Syn: Not set
        .... .... ...0 = Fin: Not set
        [TCP Flags: ·······AP···]
    Window: 245
    [Calculated window size: 31360]
    [Window size scaling factor: 128]
    Checksum: 0xa04f [unverified]
    [Checksum Status: Unverified]
    Urgent Pointer: 0
    Options: (12 bytes), No-Operation (NOP), No-Operation (NOP), Timestamps
    [SEQ/ACK analysis]
    [Timestamps]
    TCP payload (94 bytes)
Hypertext Transfer Protocol
    Data (94 bytes)

7. client libwebsockets has sent HTTP Continuation to WebSocket server
...
8. 93158	8375.154981	192.168.160.3	192.168.160.2	HTTP	169	Continuation

Issue in that libwebsockets client were sending HTTP Continuation repeatedly then received several ACK from WebSocket server and again was send HTTP Continuation so libwebsockets client can't switching to WebSocket protocol and is stuck to this HTTP Continuation state.

Could you please clarify:
1. What is reason such HTTP Continuation requests from libwebsockets client to WebSocket server?
2. How can improve fault tolerance I am subscribed to  lws_callback_function and handle LWS_CALLBACK_CLOSED and LWS_CALLBACK_CLIENT_CONNECTION_ERROR but in case HTTP Continuation libwebsockets  aren't triggered those states.

Thanks,
Igor

**Comment by lws-team** [comment]
> How can improve fault tolerance

Well, use recent lws that is supported, instead of v2.3-stable, last patched 4 years ago.

Clearly define what "fault" you are trying to "tolerate".

> What is reason such HTTP Continuation requests from libwebsockets 

I think you are getting confused by wireshark nomenclature.  There is no "http continuation".  It seems to mean that the single http transaction content takes more than one tcp packet, there is nothing wrong or strange about that.

> Issue in that libwebsockets client were sending HTTP Continuation repeatedly then received several ACK from WebSocket server and again was send HTTP Continuation so libwebsockets client can't switching to WebSocket protocol and is stuck to this HTTP Continuation state.

Let me translate that for you... you don't have any information about what your problem is.

**Comment by inline-github** [comment]
> I think you are getting confused by wireshark nomenclature. There is no "http continuation". It seems to mean that the single http transaction content takes more than one tcp packet, there is nothing wrong or strange about that.

Yes correct, I saw in pcap trace a lot of "HTTP Continuation" packets were sent to WebSocket server throughout 106 seconds then libwebsockets sent FIN packet and tried to reconnected again due to lws_client_connect_via_info called to libwebsockets . Let me rephrase my intention is detect  libwebsockets could not establish connection to WebSocket server much earlier then 106 seconds, it would be better to have 3-4 seconds timeout. Can I use for this purpose lws_set_timeout and which value enum pending_timeout should be specified ? I think in case handshake(when libwebsockets sends HTTP 405 GET ) cannot be performed PENDING_TIMEOUT_ESTABLISH_WITH_SERVER is suitable confirm please?
One more clarification if need perform reconnection after received in  lws_callback_function LWS_CALLBACK_CLOSED or LWS_CALLBACK_CLIENT_CONNECTION_ERROR  program should return -1 in order libwebsockets  can close connection then program can again call lws_client_connect_via_info to libwebsockets  correct me please.

**Comment by lws-team** [comment]
> Yes correct, I saw in pcap trace a lot of "HTTP Continuation" packets were sent to WebSocket server throughout 106 seconds then libwebsockets sent FIN packet and tried to reconnected again due to lws_client_connect_via_info called to libwebsockets .

I have no idea what is in your client or what it asks lws to do.  That could be perfectly correct.

> HTTP 405 GET

"405" has no obvious meaning there.

>  Let me rephrase my intention is detect libwebsockets could not establish connection to WebSocket server much earlier then 106 seconds,

Older lws just makes one tcp connection attempt, the OS will fail it after some OS-specific time, like 30s, if no other timeout active.  He will make one ws upgrade attempt, there are timeouts around that sequencing already, defaulting to 5s range.  He could be getting into some kind of redirect loop I guess, but even that is limited to three before it will give up.  But, eg, your code can do new connections and retries however it likes for as long as it likes.

So I don't really see whatever that problem is, is necessarily related to timeouts.  Actually, to get any clue, you need to dump the http layer requests and response packets, not at tcp layer since they're encrypted.  You can just as well dump the plaintext for them at the server as at the client.  And you need to refer to your client code.

> program should return -1 in order libwebsockets can close connection then program can again call lws_client_connect_via_info to libwebsockets correct me please.

No, what you return on those callbacks doesn't matter, the wsi is in the middle of being closed anyway.  When you return from the callback with any return value, it will complete the wsi destroy.

**Comment by inline-github** [comment]
> to get any clue, you need to dump the http layer requests and response packets, not at tcp layer since they're encrypted. You can just as well dump the plaintext for them at the server as at the client

Ok, I can dumping via tcpdump tool, Could you please more describe how possible to dump the http layer on libwebsockets client side? probably libwebsockets supports one more callback for this

**Comment by lws-team** [comment]
>  Ok, I can dumping via tcpdump tool, 

If it is using tls, tcpdump will see only encrypted data and not very useful to see what is going on.

Modern lws has handy cmake options `LWS_TLS_LOG_PLAINTEXT_RX` and `LWS_TLS_LOG_PLAINTEXT_TX` that hexdump any data going in or out respectively, just after for _RX and before for _TX the tls layer.

You can clone main and grep to see the position in the lws code they operate on, and try to hack something similar on your old version.

Only the last stable release is supported, unless it's something critical.

https://libwebsockets.org/git/libwebsockets/tree/READMEs/README.release-policy.md?h=main

**Comment by inline-github** [comment]
Ok, Thank you for detailed reply


> But, eg, your code can do new connections and retries however it likes for as long as it likes.

our WebSocket server supports only single incoming connection so need close current long(106 seconds without switching to WebSocket protocol) unsuccessful connection. As far  as I know best way to close current connection is call lws_callback_on_writable to libwebsockets then return -1 from  lws_callback_function then try to establish new connection via call lws_client_connect_via_info confirm please? does it conform to case pending( handshake in progress) connection? and does it conform  to v2.3-stable release.

**Comment by lws-team** [comment]
Yes it's a reasonable way to get the close.  In later versions you can request a wsi close next time around the event loop, but that requires more care since you have to be sure the wsi still exists.  So better to do it from inside the protocol callback.

Handshakes are protected by strict timeouts already, whatever your problem is, it shouldn't be about that.

One thing to make sure about is just because you go a WRITEABLE callback, it does imply that you could write something, but it does not imply that you should or can, or that you asked for it.  The logic of the WRITEABLE callback must understand its own state and only write if it has something to say and it knows it's the right time to say it.  A WRITEABLE callback that just always writes something, may jabber unwanted things at the wrong time.

**Comment by inline-github** [comment]
> Handshakes are protected by strict timeouts already

 Our embedded system can move system time randomly. Sorry for stupid question, only for double check. Could you please confirm libwebsockets timeouts are using monotonic time which without dependence on system time.

**Comment by lws-team** [comment]
You can check yourself what the version you are using does, since you have the sources, what it does also depends on the platform you're building for.  From memory, 4 years ago, I don't think so.

Later lws uses explicit monotonic time if the platform provides it, via `clock_gettime(CLOCK_MONOTONIC)`.  If the platform doesn't have that, it uses gettimeofday() time and provides an api to inform lws of time steps, it will adjust the expiry times accordingly.
